charms/kfp-api/src/charm.py

```diff
@@ -286,98 +202,180 @@ def set_pod_spec(self, event):
+
+        missing_attributes = [attribute for attribute in expected_attributes if attribute not in mysql]
+
+        if len(missing_attributes) == len(expected_attributes):
+            raise CheckFailed("Waiting for mysql relation data", WaitingStatus)
+        elif len(missing_attributes) > 0:
+            self.log.error(
+                f"mysql relation data missing expected attributes '{missing_attributes}'"
+            )
+            raise CheckFailed(
+                "Received incomplete data from mysql relation.  See logs", BlockedStatus
+            )
+        return mysql
+
+    def _get_object_storage(self, interfaces):
+        relation_name = "object-storage"
+        return self.validate_sdi_interface(interfaces, relation_name)
+
+    def _get_viz(self, interfaces):
+        relation_name = "kfp-viz"
+        default_viz_data = {"service-name": "unset", "service-port": "1234"}
+        return self.validate_sdi_interface(interfaces, relation_name, default_return=default_viz_data)
+
+    def validate_sdi_interface(self, interfaces: dict, relation_name: str, default_return=None):
+        """Validates data received from SerializedDataInterface, returning the data if valid
+
+        Optionally can return a default_return value when no relation is established
+
+        Raises:
+            CheckFailed(..., Blocked) when no relation established (unless default_return set)
+            CheckFailed(..., Blocked) if interface is not using SDI
+            CheckFailed(..., Blocked) if data in interface fails schema check
+            CheckFailed(..., Waiting) if we have a relation established but no data passed
+
+        Params:
+            interfaces:
+
+        Returns:
+              (dict) interface data
+        """
+        # If nothing is related to this relation, return a default value or raise an error
+        if relation_name not in interfaces or interfaces[relation_name] is None:
+            if default_return is not None:
+                return default_return
+            else:
+                raise CheckFailed(f"Missing required relation for {relation_name}", BlockedStatus)
+
+        relations = interfaces[relation_name]
+        if not isinstance(relations, SerializedDataInterface):
+            raise CheckFailed(f"Unexpected error with {relation_name} relation data - data not as expected", BlockedStatus)
+
+        # Get and validate data from the relation
+        try:
+            # relations is a dict of {(ops.model.Relation, ops.model.Application): data}
+            unpacked_relation_data = relations.get_data()
+        except ValidationError as val_error:
+            # Validation in .get_data() ensures if data is populated, it matches the schema and is
+            # not incomplete
+            self.log.error(val_error)
+            raise CheckFailed(
+                f"Found incomplete/incorrect relation data for {relation_name}.  See logs",
+                BlockedStatus
+            )
```

Comment:
This seems like something where I'm comfortable with going into an error status instead of blocked, if that happens something clearly went wrong and an error status seems a lot more appropiate than blocked.

`BlockedStatus` is for when something needs human intervention, but not for any kind of error, I would prefer if the `ValidationError` is not catched and the charm fails here.

Reply:
I was about to completely agree with this suggestion, but then thought about something like Istio with optional relations and now I'm not sure.  

Agreed we're really catching an **error** here.  If we let this raise to Juju, no further hooks fire and the charm is in stasis.  That would be fine for a required relation like here, but not an optional relation like in Istio.  I lean toward using the same pattern for both required and optional relations.  What do you think?  It is already hard to keep our many charms consistent - the fewer patterns the better imo.  Functionally, the effect of Blocked vs Error on this relation feels equivalent (they both show up loudly on `juju status` and both convey "we're broken!").  

[Juju docs](https://juju.is/docs/olm/status-values) are a bit ambiguous about the best way to do things.  My interpretation is that charms handle things they expect might happen in `Blocked`, and `Error` is for the rest.  A related charm giving me garbage data feels like something I should expect possible and handle.

Reply:
Replying to the general review comments here so I don't splinter the conversation

Re: Blocked meaning the charm worked the way it should and Error meaning the charm didn't, I agree completely.  But in the case here I think **this** charm is still working the way it should.  We can have a perfect charm, but if some other charm sends us bad data and we notice it that feels like we're blocked (given bad inputs/config) than errored

Re hiding the exceptions behind CheckFailed, agreed that could happen but I don't think it happens here.  By catching `ValidationError` we're doing a value add rather than hiding anything.  We [log the error](https://github.com/canonical/kfp-operators/pull/24/files#diff-5e034222c66adb8d88c366d84692add5e82b27063ffa982293ac8328380339d3R348) so we haven't lost anything, and we [add context saying exactly what relation broke](https://github.com/canonical/kfp-operators/pull/24/files#diff-5e034222c66adb8d88c366d84692add5e82b27063ffa982293ac8328380339d3R350).  I think by handling the error we come out ahead.  We could still catch, log context, and reraise the `ValidationError` instead of `CheckFailed`, but I think we get more this way.